crates/zkevm_circuits/src/bn254/ec_pairing/implementation.rs

```diff
@@ -0,0 +1,461 @@
+        f: &mut BN256Fq12NNField<F>,
+        line_fn: &mut LineFunctionEvaluation<F, CS>,
+    ) -> BN256Fq12NNField<F> {
+        let mut f = f.mul_by_c0c3c4(cs, &mut line_fn.c0, &mut line_fn.c3, &mut line_fn.c4);
+        f.normalize(cs);
+        f
+    }
+
+    /// Allocates the constant from `Fq2` constant
+    pub fn allocate_fq2_constant(
+        cs: &mut CS,
+        value: Fq2,
+        params: &Arc<BN256BaseNNFieldParams>,
+    ) -> BN256Fq2NNField<F> {
+        let c0 = BN256BaseNNField::allocated_constant(cs, value.c0, params);
+        let c1 = BN256BaseNNField::allocated_constant(cs, value.c1, params);
+
+        BN256Fq2NNField::new(c0, c1)
+    }
+}
+
+/// Checks the validity of the SWProjective point for the BN256 curve
+/// used in the pairing function. Namely, it verifies that the point is reduced
+/// and has a z-coordinate of one (since further the point is represented
+/// using Jacobian coordinates)
+fn validate_swprojective_point<F, CS>(
+    cs: &mut CS,
+    point: &mut BN256SWProjectivePoint<F>,
+    params: &Arc<BN256BaseNNFieldParams>,
+) where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    // Enforcing that the point is reduced
+    point.enforce_reduced(cs);
+
+    // Enforcing that the point has a z-coordinate of one
+    let mut one = BN256BaseNNField::allocated_constant(cs, BN256Fq::one(), params);
+    let mut z = point.z.clone();
+    let z_is_one = z.equals(cs, &mut one);
+    let boolean_true = Boolean::allocated_constant(cs, true);
+    Boolean::enforce_equal(cs, &z_is_one, &boolean_true);
+}
+
+/// Checks the validity of the [`BN256SWProjectivePointTwisted`]
+/// used in the pairing function. Namely, it verifies that the point is reduced
+/// and has a z-coordinate of one (since further the point is represented
+/// using Jacobian coordinates)
+fn validate_swprojective_twisted_point<F, CS>(
+    cs: &mut CS,
+    point: &mut BN256SWProjectivePointTwisted<F>,
+    params: &Arc<BN256BaseNNFieldParams>,
+) where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    // Enforcing that the point is reduced
+    point.enforce_reduced(cs);
+
+    // Enforcing that the point has z-coordinate of one
+    let mut one = BN256Fq2NNField::allocated_constant(cs, BN256Fq::one(), params);
+    let mut z = point.z.clone();
+    let z_is_one = z.equals(cs, &mut one);
+    let boolean_true = Boolean::allocated_constant(cs, true);
+    Boolean::enforce_equal(cs, &z_is_one, &boolean_true);
+}
+
+/// This function computes the pairing function for the BN256 curve using the specified method.
+pub fn ec_pairing_inner<F, CS>(
+    cs: &mut CS,
+    p: &mut BN256SWProjectivePoint<F>,
+    q: &mut BN256SWProjectivePointTwisted<F>,
+    hardexp_method: HardExpMethod,
+    compression_method: CompressionMethod,
+) -> BN256Fq12NNField<F>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    // Validating both points
+    let params = &p.x.params.clone();
+    validate_swprojective_point(cs, p, params);
```

Comment:
if point is inf then everything is broke

Reply:
The validation itself is unnecessary here because we are creating points from affine coords and z-coordinate is 1 by default. Removed it and spared the trouble.